lib/ExpressionUtils/core/exprMatrix_utils.py

```diff
@@ -203,3 +207,117 @@ def get_expression_matrix(self, params):
+        if not params.get( 'fem_object_ref' ):
+            raise ValueError( "fem_object_ref parameter not given to get_enhancedFilteredExpressionMatrix" )
+
+        fem_object_ref = params.get( 'fem_object_ref' )
+
+        fem_obj_ret = self.ws_client.get_objects2(
+                       {'objects': [{'ref': fem_object_ref }]})['data'][0]
+        fem = fem_obj_ret.get( 'data' )
+        prov = fem_obj_ret.get( 'provenance')[0]
+
+        # (1) create the enhanced FEM, starting with the FEM
+
+        efem = {}
+        for k in [ 'genome_ref', 'scale', 'type' ]:
+            efem[k] = fem.get( k )
+
+        efem['data'] = {}
+        efem['data']['col_ids'] = [ "description", 
+                                    "fold-change",
+                                    "q-value",
+                                    "min",
+                                    "max",
+                                    "mean",
+                                    "std_dev",
+                                    "is_missing_values" ]
+        efem['data']['column_labels'] =[ "Description", 
+                                         "Fold change",
+                                         "Q value",
+                                         "Min. expression",
+                                         "Max. expression",
+                                         "Mean expression",
+                                         "Std. dev.",
+                                         "Missing values?" ]
+        fm = fem.get('data')
+        efem['data']['row_ids'] = fm.get('row_ids')
+        efem['data']['values' ] = []
+        n_efem_rows = len( efem['data']['row_ids'] )
+        fvals = fm.get('values')
+        if ( len( fvals ) != n_efem_rows ):
+            raise Exception( "length discrepancy in filtered expression matrix: {0} row_ids but {1} values".format( n_efem_rows, len( fvals ) ) )
+
+        for i in range( 0, n_efem_rows ):
+            efem['data']['values'].append( [ 'NA', 'NA', 'NA' ] + self.get_matrix_stats( fvals[i] ) )
+
+        # (2) Get genome object and feature descriptions, put those in column 1
+
+        feat_dict = self.gaa.get_feature_functions( { 'ref': fem.get( 'genome_ref' ), 'feature_id_list': None } )
+        for i in range( 0, n_efem_rows ):
+            desc = feat_dict.get( efem['data']['row_ids'][i] )
+            if desc:
+                efem['data']['values'][i][0] = desc     # leave as 'NA' if no entry in feat_dict
+
+        # (3) get DEM from provenance and merge the FC and q-values
+
+        if prov.get( 'input_ws_objects' ):
+            dem_ref = prov.get( 'input_ws_objects' )[0]
+            dem_obj_ret = self.ws_client.get_objects2(
+                           {'objects': [{'ref': dem_ref }]})['data'][0]
+            dem = dem_obj_ret.get( 'data' )
+            
+            dem_dict = self.convert_dem_to_dict( dem.get('data') )  # convert to dictionary for quick lookups
+
+            for i in range( 0, n_efem_rows ):
```

Comment:
This is your third pass over the rows in the data. It would be more efficient ad a bit clearer to gather the dem_dict, feature_dict and matrix stats at the top the traverse the rows once